security/s2a/internal/crypter/halfconn.go

```diff
@@ -0,0 +1,126 @@
+	cs := NewCiphersuite(ciphersuite)
+	if cs.trafficSecretSize() != len(trafficSecret) {
+		return S2AHalfConnection{}, fmt.Errorf("supplied traffic secret must be %v bytes, given: %v", cs.trafficSecretSize(), trafficSecret)
+	}
+
+	hc := S2AHalfConnection{h: cs.hashFunction(), expander: &defaultHKDFExpander{}, seqCounter: newCounter()}
+	key, err := hc.deriveSecret(hc.h, trafficSecret, []byte(tls13Key))
+	if err != nil {
+		return S2AHalfConnection{}, fmt.Errorf("hc.deriveSecret(h, %v, %v) failed with error: %v", trafficSecret, tls13Key, err)
+	}
+	hc.aeadCrypter, err = cs.aeadCrypter(key)
+	if err != nil {
+		return S2AHalfConnection{}, fmt.Errorf("cs.aeadCrypter(%v) failed with error: %v", key, err)
+	}
+	hc.nonce, err = hc.deriveSecret(hc.h, trafficSecret, []byte(tls13Nonce))
+	if err != nil {
+		return S2AHalfConnection{}, fmt.Errorf("hc.deriveSecret(h, %v, %v) failed with error: %v", trafficSecret, tls13Nonce, err)
+	}
+	return hc, nil
+}
+
+func (hc *S2AHalfConnection) Encrypt(dst, plaintext, aad []byte) ([]byte, error) {
+	hc.mutex.Lock()
+	sequence, err := hc.getAndIncrementSequence()
+	if err != nil {
+		hc.mutex.Unlock()
+		return nil, err
+	}
+	nonce := hc.maskedNonce(sequence)
+	crypter := hc.aeadCrypter
+	hc.mutex.Unlock()
+	return crypter.encrypt(dst, plaintext, nonce, aad)
+}
+
+func (hc *S2AHalfConnection) Decrypt(dst, ciphertext, aad []byte) ([]byte, error) {
+	hc.mutex.Lock()
+	sequence, err := hc.getAndIncrementSequence()
+	if err != nil {
+		hc.mutex.Unlock()
+		return nil, err
+	}
+	nonce := hc.maskedNonce(sequence)
+	crypter := hc.aeadCrypter
+	hc.mutex.Unlock()
+	return crypter.decrypt(dst, ciphertext, nonce, aad)
+}
+
+func (hc *S2AHalfConnection) UpdateKey() error {
+	var err error
+	hc.trafficSecret, err = hc.deriveSecret(hc.h, hc.trafficSecret, []byte(tls13Update))
+	if err != nil {
+		return fmt.Errorf("hc.deriveSecret(h, %v, %v) failed with error: %v", hc.trafficSecret, tls13Update, err)
+	}
+	key, err := hc.deriveSecret(hc.h, hc.trafficSecret, []byte(tls13Key))
+	if err != nil {
+		return fmt.Errorf("hc.deriveSecret(h, %v, %v) failed with error: %v", hc.trafficSecret, tls13Key, err)
+	}
+	err = hc.aeadCrypter.updateKey(key)
+	if err != nil {
+		return fmt.Errorf("hc.aeadCrypter.updateKey(%v) failed with error: %v", key, err)
+	}
+	hc.nonce, err = hc.deriveSecret(hc.h, hc.trafficSecret, []byte(tls13Nonce))
+	if err != nil {
+		return fmt.Errorf("hc.deriveSecret(h, %v, %v) failed with error: %v", hc.trafficSecret, tls13Nonce, err)
+	}
+	hc.seqCounter.reset()
+	return nil
+}
+
+func (hc *S2AHalfConnection) getAndIncrementSequence() (uint64, error) {
+	sequence, err := hc.seqCounter.val()
+	if err != nil {
+		return 0, err
+	}
+	hc.seqCounter.inc()
+	return sequence, nil
+}
+
+func (hc *S2AHalfConnection) maskedNonce(sequence uint64) []byte {
+	for i := 0; i < uint64Size; i++ {
+		hc.nonce[nonceSize-8+i] ^= sequence >> (56 - 8*i)
```

Comment:
Could you add an extra set of parentheses around `sequence >> (56-8*i)`? Doing so will make this a bit more readable, I think.

Reply:
The linter complains about unnecessary parenthesis so I think it's best that we follow what the linter suggests.

Reply:
Ah ok. Let's follow the linter then. :)